File: frontend/src/services/storage.ts
import { apiFetch } from './client';

export interface StorageOverview {
  connected: boolean;
  pvcCount: number;
  boundPvcCount: number;
  pendingPvcCount: number;
  pvCount: number;
  availablePvCount: number;
  storageClassCount: number;
  totalRequested: string;
  totalCapacity: string;
  totalRequestedBytes: number;
  totalCapacityBytes: number;
}

export interface PersistentVolumeClaimInfo {
  name: string;
  namespace: string;
  phase: string;
  volumeName: string;
  requested: string;
  capacity: string;
  requestedBytes: number;
  capacityBytes: number;
  storageClass: string;
  accessModes: string[];
  volumeMode: string;
  createdAt: string;
  usedBy: string[];
}

export interface PersistentVolumeInfo {
  name: string;
  phase: string;
  capacity: string;
  capacityBytes: number;
  storageClass: string;
  accessModes: string[];
  reclaimPolicy: string;
  claim: string;
  driver: string;
  createdAt: string;
}

export interface StorageClassInfo {
  name: string;
  provisioner: string;
  reclaimPolicy: string;
  volumeBindingMode: string;
  allowVolumeExpansion: boolean;
  isDefault: boolean;
  createdAt: string;
}

export interface NodeStorageInfo {
  name: string;
  containerRuntime: string;
  runtimeName: string;
  runtimeVersion: string;
  kubeletVersion: string;
  osImage: string;
  kernelVersion: string;
  architecture: string;
  operatingSystem: string;
  ephemeralStorageCapacity: string;
  ephemeralStorageAllocatable: string;
  ephemeralStorageCapacityBytes: number;
  ephemeralStorageAllocatableBytes: number;
  imageCount: number;
  imageBytes: number;
  imageSize: string;
  diskPressure: boolean;
  ready: boolean;
}

/**
 * protojson serialises 64-bit integers as JSON *strings* to avoid the precision
 * loss of an IEEE-754 double. Every byte count below therefore arrives as
 * "10737418240" rather than a number, and arithmetic on it would concatenate
 * instead of adding.
 */
function toNumber(value: unknown): number {
  if (typeof value === 'number') return value;
  if (typeof value === 'string' && value !== '') {
    const parsed = Number(value);
    return Number.isFinite(parsed) ? parsed : 0;
  }
  return 0;
}

async function post(procedure: string, body: Record<string, unknown> = {}): Promise<any> {
  const response = await apiFetch(`/idp.v1.StorageService/${procedure}`, body);
  if (!response.ok) {
    // Connect returns a JSON body carrying the human-readable reason; the bare
    // status text alone ("Bad Request") tells the user nothing actionable.
    const detail = await response.json().catch(() => null);
    throw new Error(detail?.message || `${procedure} failed: ${response.statusText}`);
  }
  return response.json();
}

export async function getStorageOverview(): Promise<StorageOverview> {
  const data = await post('GetStorageOverview');
  return {
    connected: data.connected ?? false,
    pvcCount: data.pvcCount ?? 0,
    boundPvcCount: data.boundPvcCount ?? 0,
    pendingPvcCount: data.pendingPvcCount ?? 0,
    pvCount: data.pvCount ?? 0,
    availablePvCount: data.availablePvCount ?? 0,
    storageClassCount: data.storageClassCount ?? 0,
    totalRequested: data.totalRequested || '0',
    totalCapacity: data.totalCapacity || '0',
    totalRequestedBytes: toNumber(data.totalRequestedBytes),
    totalCapacityBytes: toNumber(data.totalCapacityBytes),
  };
}

export async function listPersistentVolumeClaims(
  namespace = '',
): Promise<PersistentVolumeClaimInfo[]> {
  const data = await post('ListPersistentVolumeClaims', { namespace });
  return (data.claims ?? []).map((c: any) => ({
    name: c.name ?? '',
    namespace: c.namespace ?? '',
    phase: c.phase || 'Unknown',
    volumeName: c.volumeName ?? '',
    requested: c.requested ?? '',
    capacity: c.capacity ?? '',
    requestedBytes: toNumber(c.requestedBytes),
    capacityBytes: toNumber(c.capacityBytes),
    storageClass: c.storageClass ?? '',
    accessModes: c.accessModes ?? [],
    volumeMode: c.volumeMode ?? '',
    createdAt: c.createdAt ?? '',
    usedBy: c.usedBy ?? [],
  }));
}

export async function listPersistentVolumes(): Promise<PersistentVolumeInfo[]> {
  const data = await post('ListPersistentVolumes');
  return (data.volumes ?? []).map((v: any) => ({
    name: v.name ?? '',
    phase: v.phase || 'Unknown',
    capacity: v.capacity ?? '',
    capacityBytes: toNumber(v.capacityBytes),
    storageClass: v.storageClass ?? '',
    accessModes: v.accessModes ?? [],
    reclaimPolicy: v.reclaimPolicy ?? '',
    claim: v.claim ?? '',
    driver: v.driver ?? '',
    createdAt: v.createdAt ?? '',
  }));
}

export async function listStorageClasses(): Promise<StorageClassInfo[]> {
  const data = await post('ListStorageClasses');
  return (data.storageClasses ?? []).map((s: any) => ({
    name: s.name ?? '',
    provisioner: s.provisioner ?? '',
    reclaimPolicy: s.reclaimPolicy ?? '',
    volumeBindingMode: s.volumeBindingMode ?? '',
    allowVolumeExpansion: s.allowVolumeExpansion ?? false,
    isDefault: s.isDefault ?? false,
    createdAt: s.createdAt ?? '',
  }));
}

export async function listNodeStorage(): Promise<NodeStorageInfo[]> {
  const data = await post('ListNodeStorage');
  return (data.nodes ?? []).map((n: any) => ({
    name: n.name ?? '',
    containerRuntime: n.containerRuntime ?? '',
    runtimeName: n.runtimeName ?? '',
    runtimeVersion: n.runtimeVersion ?? '',
    kubeletVersion: n.kubeletVersion ?? '',
    osImage: n.osImage ?? '',
    kernelVersion: n.kernelVersion ?? '',
    architecture: n.architecture ?? '',
    operatingSystem: n.operatingSystem ?? '',
    ephemeralStorageCapacity: n.ephemeralStorageCapacity ?? '',
    ephemeralStorageAllocatable: n.ephemeralStorageAllocatable ?? '',
    ephemeralStorageCapacityBytes: toNumber(n.ephemeralStorageCapacityBytes),
    ephemeralStorageAllocatableBytes: toNumber(n.ephemeralStorageAllocatableBytes),
    imageCount: n.imageCount ?? 0,
    imageBytes: toNumber(n.imageBytes),
    imageSize: n.imageSize || '0',
    diskPressure: n.diskPressure ?? false,
    ready: n.ready ?? false,
  }));
}

File: backend/internal/kubernetes/storage.go
package kubernetes

import (
	"context"
	"fmt"
	"sort"
	"strings"
	"time"

	corev1 "k8s.io/api/core/v1"
	"k8s.io/apimachinery/pkg/api/resource"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

// isDefaultStorageClassAnnotation marks the class the API server falls back to
// when a claim names none. The beta key is still what most clusters carry.
const (
	isDefaultStorageClassAnnotation     = "storageclass.kubernetes.io/is-default-class"
	isDefaultStorageClassAnnotationBeta = "storageclass.beta.kubernetes.io/is-default-class"
)

// PersistentVolumeClaim is the read model for one claim.
type PersistentVolumeClaim struct {
	Name           string
	Namespace      string
	Phase          string
	VolumeName     string
	Requested      string
	Capacity       string
	RequestedBytes int64
	CapacityBytes  int64
	StorageClass   string
	AccessModes    []string
	VolumeMode     string
	CreatedAt      string
	UsedBy         []string
}

// PersistentVolume is the read model for one cluster volume.
type PersistentVolume struct {
	Name          string
	Phase         string
	Capacity      string
	CapacityBytes int64
	StorageClass  string
	AccessModes   []string
	ReclaimPolicy string
	Claim         string
	Driver        string
	CreatedAt     string
}

// StorageClass is the read model for one provisioner class.
type StorageClass struct {
	Name                 string
	Provisioner          string
	ReclaimPolicy        string
	VolumeBindingMode    string
	AllowVolumeExpansion bool
	IsDefault            bool
	CreatedAt            string
}

// NodeStorage carries a node's container runtime and local disk footprint.
type NodeStorage struct {
	Name                             string
	ContainerRuntime                 string
	RuntimeName                      string
	RuntimeVersion                   string
	KubeletVersion                   string
	OSImage                          string
	KernelVersion                    string
	Architecture                     string
	OperatingSystem                  string
	EphemeralStorageCapacity         string
	EphemeralStorageAllocatable      string
	EphemeralStorageCapacityBytes    int64
	EphemeralStorageAllocatableBytes int64
	ImageCount                       int32
	ImageBytes                       int64
	ImageSize                        string
	DiskPressure                     bool
	Ready                            bool
}

// StorageOverview aggregates the counts a dashboard leads with.
type StorageOverview struct {
	PVCCount            int32
	BoundPVCCount       int32
	PendingPVCCount     int32
	PVCount             int32
	AvailablePVCount    int32
	StorageClassCount   int32
	TotalRequested      string
	TotalCapacity       string
	TotalRequestedBytes int64
	TotalCapacityBytes  int64
}

// ListPersistentVolumeClaims returns claims in a namespace, or across all
// namespaces when namespace is empty.
//
// Pods are listed alongside so each claim can report what is mounting it: a
// claim nothing references is still consuming its volume, and that is invisible
// from the claim object alone.
func (c *Client) ListPersistentVolumeClaims(ctx context.Context, namespace string) ([]PersistentVolumeClaim, error) {
	cs, csErr := c.cs()
	if csErr != nil {
		return nil, csErr
	}
	list, err := cs.CoreV1().PersistentVolumeClaims(namespace).List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("list persistentvolumeclaims: %w", err)
	}

	// A failure here degrades the "used by" column to empty rather than
	// failing the whole page — the claims themselves are still worth showing.
	mounts := map[string][]string{}
	if pods, podErr := cs.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{}); podErr == nil {
		for i := range pods.Items {
			pod := &pods.Items[i]
			for _, vol := range pod.Spec.Volumes {
				if vol.PersistentVolumeClaim == nil {
					continue
				}
				key := pod.Namespace + "/" + vol.PersistentVolumeClaim.ClaimName
				mounts[key] = append(mounts[key], pod.Name)
			}
		}
	}

	claims := make([]PersistentVolumeClaim, 0, len(list.Items))
	for i := range list.Items {
		pvc := &list.Items[i]

		requested, requestedBytes := quantityFrom(pvc.Spec.Resources.Requests, corev1.ResourceStorage)
		capacity, capacityBytes := quantityFrom(pvc.Status.Capacity, corev1.ResourceStorage)

		volumeMode := ""
		if pvc.Spec.VolumeMode != nil {
			volumeMode = string(*pvc.Spec.VolumeMode)
		}

		storageClass := ""
		if pvc.Spec.StorageClassName != nil {
			storageClass = *pvc.Spec.StorageClassName
		}

		usedBy := mounts[pvc.Namespace+"/"+pvc.Name]
		sort.Strings(usedBy)

		claims = append(claims, PersistentVolumeClaim{
			Name:           pvc.Name,
			Namespace:      pvc.Namespace,
			Phase:          string(pvc.Status.Phase),
			VolumeName:     pvc.Spec.VolumeName,
			Requested:      requested,
			Capacity:       capacity,
			RequestedBytes: requestedBytes,
			CapacityBytes:  capacityBytes,
			StorageClass:   storageClass,
			AccessModes:    accessModeStrings(pvc.Spec.AccessModes),
			VolumeMode:     volumeMode,
			CreatedAt:      formatTimestamp(pvc.CreationTimestamp),
			UsedBy:         usedBy,
		})
	}

	sort.Slice(claims, func(a, b int) bool {
		if claims[a].Namespace != claims[b].Namespace {
			return claims[a].Namespace < claims[b].Namespace
		}
		return claims[a].Name < claims[b].Name
	})

	return claims, nil
}

// ListPersistentVolumes returns every cluster-scoped volume.
func (c *Client) ListPersistentVolumes(ctx context.Context) ([]PersistentVolume, error) {
	cs, csErr := c.cs()
	if csErr != nil {
		return nil, csErr
	}
	list, err := cs.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("list persistentvolumes: %w", err)
	}

	volumes := make([]PersistentVolume, 0, len(list.Items))
	for i := range list.Items {
		pv := &list.Items[i]

		capacity, capacityBytes := quantityFrom(pv.Spec.Capacity, corev1.ResourceStorage)

		claim := ""
		if pv.Spec.ClaimRef != nil {
			claim = pv.Spec.ClaimRef.Namespace + "/" + pv.Spec.ClaimRef.Name
		}

		volumes = append(volumes, PersistentVolume{
			Name:          pv.Name,
			Phase:         string(pv.Status.Phase),
			Capacity:      capacity,
			CapacityBytes: capacityBytes,
			StorageClass:  pv.Spec.StorageClassName,
			AccessModes:   accessModeStrings(pv.Spec.AccessModes),
			ReclaimPolicy: string(pv.Spec.PersistentVolumeReclaimPolicy),
			Claim:         claim,
			Driver:        volumeDriver(pv),
			CreatedAt:     formatTimestamp(pv.CreationTimestamp),
		})
	}

	sort.Slice(volumes, func(a, b int) bool { return volumes[a].Name < volumes[b].Name })
	return volumes, nil
}

// ListStorageClasses returns the provisioner classes available to claims.
func (c *Client) ListStorageClasses(ctx context.Context) ([]StorageClass, error) {
	cs, csErr := c.cs()
	if csErr != nil {
		return nil, csErr
	}
	list, err := cs.StorageV1().StorageClasses().List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("list storageclasses: %w", err)
	}

	classes := make([]StorageClass, 0, len(list.Items))
	for i := range list.Items {
		sc := &list.Items[i]

		reclaim := ""
		if sc.ReclaimPolicy != nil {
			reclaim = string(*sc.ReclaimPolicy)
		}
		binding := ""
		if sc.VolumeBindingMode != nil {
			binding = string(*sc.VolumeBindingMode)
		}
		expansion := sc.AllowVolumeExpansion != nil && *sc.AllowVolumeExpansion

		classes = append(classes, StorageClass{
			Name:                 sc.Name,
			Provisioner:          sc.Provisioner,
			ReclaimPolicy:        reclaim,
			VolumeBindingMode:    binding,
			AllowVolumeExpansion: expansion,
			IsDefault: sc.Annotations[isDefaultStorageClassAnnotation] == "true" ||
				sc.Annotations[isDefaultStorageClassAnnotationBeta] == "true",
			CreatedAt: formatTimestamp(sc.CreationTimestamp),
		})
	}

	sort.Slice(classes, func(a, b int) bool { return classes[a].Name < classes[b].Name })
	return classes, nil
}

// ListNodeStorage reports each node's container runtime and image footprint.
func (c *Client) ListNodeStorage(ctx context.Context) ([]NodeStorage, error) {
	cs, csErr := c.cs()
	if csErr != nil {
		return nil, csErr
	}
	list, err := cs.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
	if err != nil {
		return nil, fmt.Errorf("list nodes: %w", err)
	}

	nodes := make([]NodeStorage, 0, len(list.Items))
	for i := range list.Items {
		node := &list.Items[i]
		info := node.Status.NodeInfo

		runtimeName, runtimeVersion := splitRuntime(info.ContainerRuntimeVersion)

		capacity, capacityBytes := quantityFrom(node.Status.Capacity, corev1.ResourceEphemeralStorage)
		allocatable, allocatableBytes := quantityFrom(node.Status.Allocatable, corev1.ResourceEphemeralStorage)

		var imageBytes int64
		for _, img := range node.Status.Images {
			imageBytes += img.SizeBytes
		}

		var diskPressure, ready bool
		for _, cond := range node.Status.Conditions {
			switch cond.Type {
			case corev1.NodeDiskPressure:
				diskPressure = cond.Status == corev1.ConditionTrue
			case corev1.NodeReady:
				ready = cond.Status == corev1.ConditionTrue
			}
		}

		nodes = append(nodes, NodeStorage{
			Name:                             node.Name,
			ContainerRuntime:                 info.ContainerRuntimeVersion,
			RuntimeName:                      runtimeName,
			RuntimeVersion:                   runtimeVersion,
			KubeletVersion:                   info.KubeletVersion,
			OSImage:                          info.OSImage,
			KernelVersion:                    info.KernelVersion,
			Architecture:                     info.Architecture,
			OperatingSystem:                  info.OperatingSystem,
			EphemeralStorageCapacity:         capacity,
			EphemeralStorageAllocatable:      allocatable,
			EphemeralStorageCapacityBytes:    capacityBytes,
			EphemeralStorageAllocatableBytes: allocatableBytes,
			ImageCount:                       int32(len(node.Status.Images)),
			ImageBytes:                       imageBytes,
			ImageSize:                        FormatBytes(imageBytes),
			DiskPressure:                     diskPressure,
			Ready:                            ready,
		})
	}

	sort.Slice(nodes, func(a, b int) bool { return nodes[a].Name < nodes[b].Name })
	return nodes, nil
}

// GetStorageOverview aggregates claim, volume and class counts.
func (c *Client) GetStorageOverview(ctx context.Context) (StorageOverview, error) {
	claims, err := c.ListPersistentVolumeClaims(ctx, "")
	if err != nil {
		return StorageOverview{}, err
	}

	volumes, err := c.ListPersistentVolumes(ctx)
	if err != nil {
		return StorageOverview{}, err
	}

	classes, err := c.ListStorageClasses(ctx)
	if err != nil {
		return StorageOverview{}, err
	}

	overview := StorageOverview{
		PVCCount:          int32(len(claims)),
		PVCount:           int32(len(volumes)),
		StorageClassCount: int32(len(classes)),
	}

	for _, claim := range claims {
		switch claim.Phase {
		case string(corev1.ClaimBound):
			overview.BoundPVCCount++
		case string(corev1.ClaimPending):
			overview.PendingPVCCount++
		}
		overview.TotalRequestedBytes += claim.RequestedBytes
	}

	for _, volume := range volumes {
		if volume.Phase == string(corev1.VolumeAvailable) {
			overview.AvailablePVCount++
		}
		overview.TotalCapacityBytes += volume.CapacityBytes
	}

	overview.TotalRequested = FormatBytes(overview.TotalRequestedBytes)
	overview.TotalCapacity = FormatBytes(overview.TotalCapacityBytes)

	return overview, nil
}

// quantityFrom reads one resource out of a quantity map, returning both the
// canonical string and the byte count. A missing key yields ("", 0) rather than
// "0" so the UI can distinguish "unset" from "zero".
func quantityFrom(list corev1.ResourceList, name corev1.ResourceName) (string, int64) {
	qty, ok := list[name]
	if !ok {
		return "", 0
	}
	return qty.String(), qty.Value()
}

// FormatBytes renders a byte count in Kubernetes binary notation (Ki/Mi/Gi),
// which is what users see in manifests and in kubectl output.
func FormatBytes(bytes int64) string {
	if bytes <= 0 {
		return "0"
	}
	return resource.NewQuantity(bytes, resource.BinarySI).String()
}

func accessModeStrings(modes []corev1.PersistentVolumeAccessMode) []string {
	out := make([]string, 0, len(modes))
	for _, mode := range modes {
		out = append(out, string(mode))
	}
	return out
}

// splitRuntime turns "containerd://1.7.13" into ("containerd", "1.7.13").
func splitRuntime(runtime string) (string, string) {
	name, version, found := strings.Cut(runtime, "://")
	if !found {
		return runtime, ""
	}
	return name, version
}

// volumeDriver names whatever backs the volume: the CSI driver where one is in
// play, otherwise the in-tree source kind.
func volumeDriver(pv *corev1.PersistentVolume) string {
	switch {
	case pv.Spec.CSI != nil:
		return pv.Spec.CSI.Driver
	case pv.Spec.HostPath != nil:
		return "hostPath"
	case pv.Spec.Local != nil:
		return "local"
	case pv.Spec.NFS != nil:
		return "nfs"
	default:
		return "unknown"
	}
}

func formatTimestamp(ts metav1.Time) string {
	if ts.IsZero() {
		return ""
	}
	return ts.UTC().Format(time.RFC3339)
}
